docs/proposals/cluster-autoscaling.md

```diff
@@ -0,0 +1,269 @@
+                 |              |          |           |         |
+            +----v----+         |      +---v----+      |         |
+            |   KAC   |<-+      +----> |Heapster| <----+         |
+            +---------+  |             +---+----+                |
+                 |       |                 |                     |
+                 |       +-----------------+                     |
+ Kubernetes      |                                               |
+-----------------|-----------------------------------------------|------
+ Cloud provider  |                                               |
+                 |                                               |
+                 |                                       +-------+-------+
+                 +-------------------------------------> |Node controller|
+                                                         +---------------+
+```
+
+This will solve the following problems:
+(I2) [GCE specific logic] As cluster autoscaling will have most of its components inside the cluster it will be provider independent (I2).
+
+At this point we will encourage the community to contribute code that talks to particular cloud provider node controller.
+
+## Kubernetes 1.4-1.5 - big KAC
+
+Assuming the need for more sophisticated cluster autoscaling we will provide a full blown autoscaler that will solve (I3)[Works only for well formed pods] and will:
+
+* Take into account the number of pending pods and will check whether adding a node will allow to schedule them.
+* Choose the best node to kill once the downscale decision was made (and before will make sure that the homeless pods will schedule somewhere).
+* Reasonably handle batch-like best-effort workload
+* Use rescheduler to redistribute the load (https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/rescheduler.md).
+
+All of that will happen inside of the Kubernetes cluster. The exact details are to be decided.
+
+
+## Appendix
+
+## Small task scheduling guarantee
+
+Each pod may provide CPU request information - minimal amount of cpu that is required for pod to run.
+It may consume more, if available but it will have this amount guaranteed. On a machine with 1 CPU we can
+put pods that request for 0.2, 0.3, 0.4 CPU respectively. The total amount of requested CPU will be then 0.9.
+
+In autoscaler we can specify that nodes should have pods with total amount of X CPU, on average.
+If this average is higher, a new node is created and new pods will most likely go there. Average of
+X CPU guarantees that there there is a node whose CPU is committed to less or equal to X. Thus there is
+always a node that can fit a pod with request set to `1-X`.
+
+If a pod has some bigger request then there is no guarantee. It may be a problem for customers who would
+like to have big pods running on nodes or even devote nodes to one type of job (some kind of database for instance).
```

Comment:
And:
* pods that use hostPort
* pods with additional constraints, such as nodeName or nodeSelector
* pods that use hostPath and expect the data to remain resident